Remove formatting from ListGroupsService and make api_groups view use JSON presenter

lyzadanger opened the pull request:
This PR is a chunk of #4785 

This PR adapts `ListGroupsService` to remove legacy formatting of groups objects—`ListGroupsService` now deals exclusively in `Group` model objects. `api_groups` view now uses the `GroupsJSONPresenter` to format groups it gets from the `ListGroupsService.`

Also took the opportunity to refactor `GroupsJSONPresenter` not to inherit from `GroupJSONPresenter` but instead utilize `GroupJSONPresenter` instances in its `asdict` method.

Once _this_ PR is done and dusted, I can submit a PR to make `session` use all this new stuff and then finally we can kill `AuthorityGroupService` and `ProfileGroupService`

seanh commented:
It's not something to do as part of this work but I wonder what would happen if we went through all the presenter classes replacing `asdict()` with `asdicts()`? How much would that simplify and improve the code that uses these classes and especially the tests for that user code?

seanh commented on tests/h/views/api_groups_test.py:
```diff
@@ -12,44 +12,65 @@
-        pyramid_request.user = user
-        pyramid_request.json_body = {}
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.route_url = mock.Mock(return_value='/groups/foo')
```
Ideally we should use `routes` fixtures instead of mocking `route_url`.

lyzadanger replied:
Hi! Can you elaborate a bit here? I think I see how I could extend the `routes` fixture (which I finally located in `tests/h/views/main_test`...I think?) but I'm not sure how to apply that to `route_url` the method.

seanh replied:
Ah no, there's no one routes fixture, rather routes fixtures are a pattern used throughout the tests. You'll find lots of fixtures called `routes` defined in different test classes and modules and used within those class / modules. So you could define your own routes fixture in this class. Something like:

```
    @pytest.fixture
    def routes(self, pyramid_config):
        pyramid_config.add_route('group_read', '/groups/{pubid}/{slug}')
```

It's a fixture that modifies the `pyramid_config` fixture that in turn modifies the `pyramid_request` fixture! So the end result is that the dummy `pyramid_request` receives the route definition.

lyzadanger replied:
But does `pyramid_request` have a `route_url` method? I'm unclear of how to mock that such that it interfaces with the `routes` fixture.

seanh replied:
Wait a sec, `api_groups.py` doesn't call `route_url`! It just passes it in to the (mock) `GroupJSONPresenter`. So this isn't needed at all - the `pyramid_request` fixture function here can just be deleted.

For the record though `pyramid_request` _does_ have a `route_url()` method. It's not a mock object, it's a [DummyRequest](https://docs.pylonsproject.org/projects/pyramid//en/latest/api/testing.html?highlight=dummyrequest#pyramid.testing.DummyRequest). Our `pyramid_request` fixture just wraps Pyramid's `DummyRequest` class making it available as a pytest fixture:

https://github.com/hypothesis/h/blob/74c60e9f7c15222731cc583af7728bd039a56dd0/tests/h/conftest.py#L227-L239

If you create a `routes` fixture that uses the `pyramid_config` fixture and calls `pyramid_config.add_route()` that route that you added ends up on the dummy `pyramid_request` object because our `pyramid_config` fixture applies itself to the `pyramid_request` fixture:

https://github.com/hypothesis/h/blob/74c60e9f7c15222731cc583af7728bd039a56dd0/tests/h/conftest.py#L215-L224

(Yes, fixtures can mess with other fixtures.)

So if `api_groups.py` were actually calling `route_url()` this would work:

```diff
diff --git a/tests/h/views/api_groups_test.py b/tests/h/views/api_groups_test.py
index 66374e62c..66d8c0d04 100644
--- a/tests/h/views/api_groups_test.py
+++ b/tests/h/views/api_groups_test.py
@@ -14,6 +14,7 @@ from h.services.group import GroupService
 pytestmark = pytest.mark.usefixtures('GroupsJSONPresenter')
 
 
+@pytest.mark.usefixtures('routes')
 class TestGroups(object):
 
     def test_all_groups_proxies_to_service(self, anonymous_request, list_groups_service):
@@ -60,9 +61,8 @@ class TestGroups(object):
         assert result == GroupsJSONPresenter(open_groups, anonymous_request).asdicts.return_value
 
     @pytest.fixture
-    def pyramid_request(self, pyramid_request):
-        pyramid_request.route_url = mock.Mock(return_value='/groups/foo')
-        return pyramid_request
+    def routes(self, pyramid_config):
+        pyramid_config.add_route('group_read', '/groups/{pubid}')
 
     @pytest.fixture
     def open_groups(self, factories):
```

lyzadanger commented:
@seanh Good gravy, those tests were gross. Thank you for helping me see the error in my ways! I've addressed your comments with the exception of the renaming-of-patched-presenter-fixture and use-routes-fixture as commented previously.

I agree with you that the presenters could use some refactoring to account for collections throughout. Let's hold that thought?

Could you take another peek?

seanh commented on tests/h/presenters/group_json_test.py:
```diff
@@ -81,3 +99,8 @@ def test_asdicts_injects_urls(self, factories):
+
+
+@pytest.fixture
+def group_json_presenter(patch):
```
Oh I see, yeah that's unfortunate! Just `GroupJSONPresenter_` with a trailing underscore maybe, sucks but that's the typical Python technique for just avoiding a naming conflict

seanh replied:
_Or_, since I guess the `group_json_presenter` fixture is only used in the `TestGroupsJSONPresenter` class, if you define the fixture in the test class instead of at module level then it will override the conflicting name, might work

seanh commented:
Looks good! I added new suggestions for the routes fixture and the `GroupJSONPresenter` fixture name collision. Do you want to try those?

lyzadanger commented:
OK, I think we're down to just that `routes` fixture thang.

lyzadanger commented:
I've addressed the last item in @seanh 's feedback, so I'm going to ask forgiveness later but merge now.
